model/src/main/java/io/spine/validation/RequiredRule.java

```diff
@@ -48,32 +53,89 @@ private RequiredRule() {
+            return Optional.empty();
+        }
+        SimpleRule integratedRule = rule(
+                field, unsetValue.get(), errorMessage, "Field must be set.", false
+        );
+        if (!isRepeated(field)) {
+            return Optional.of(wrap(integratedRule));
+        }
+        Optional<Value> singularUnsetValue = UnsetValue.singular(field.getType());
+        if (!singularUnsetValue.isPresent()) {
+            return Optional.of(wrap(integratedRule));
+        }
+        SimpleRule differentialRule = rule(
+                field, singularUnsetValue.get(), errorMessage, "", true
+        );
+        CompositeRule composite = collectionRule(field, integratedRule, differentialRule);
+        return Optional.of(wrap(composite));
+    }
+
+    private static CompositeRule collectionRule(Field field,
+                                                SimpleRule integratedRule,
+                                                SimpleRule differentialRule) {
+        @SuppressWarnings("DuplicateStringLiteralInspection")
+        CompositeRule composite = CompositeRule.newBuilder()
+                .setLeft(wrap(integratedRule))
+                .setOperator(AND)
+                .setRight(wrap(differentialRule))
+                .setErrorMessage("Collection must not be empty and cannot contain default values.")
+                .setField(field.getName())
+                .build();
+        return composite;
+    }
+
+    private static SimpleRule rule(Field field,
+                                   Value value,
+                                   String errorMessage,
+                                   String defaultErrorMessage,
+                                   boolean distibute) {
+        String msg = errorMessage.isEmpty() ? defaultErrorMessage : errorMessage;
+        return SimpleRule
                 .newBuilder()
-                .setErrorMessage(errorMessage)
+                .setErrorMessage(msg)
                 .setField(field.getName())
                 .setOperator(NOT_EQUAL)
-                .setOtherValue(unsetValue)
-                .setDistribute(false)
+                .setOtherValue(value)
+                .setDistribute(distibute)
                 .vBuild();
-        return rule;
     }
 
-    private static IllegalStateException doesNotSupportRequired(Field field) {
-        String fieldName = field.getName()
-                                .getValue();
-        String typeUrl = typeUrl(field.getDeclaringType());
-        PrimitiveType type = field.getType()
-                                  .getPrimitive();
-        return new IllegalStateException(format(
-                "Field `%s.%s` of type `%s` does not support `(required)` validation.",
-                typeUrl, fieldName, type
-        ));
+    private static Rule wrap(SimpleRule r) {
+        return Rule.newBuilder()
+                .setSimple(r)
+                .build();
+    }
+
+    private static Rule wrap(CompositeRule r) {
```

Comment:
Can these two methods be a part of a mixin interface common for `SimpleRule` and `CompositeRule`?